dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/export/changelog/ChangeLogServiceTest.java

```diff
@@ -0,0 +1,311 @@
+import org.hisp.dhis.program.Event;
+import org.hisp.dhis.tracker.TrackerTest;
+import org.hisp.dhis.tracker.imports.TrackerImportParams;
+import org.hisp.dhis.tracker.imports.TrackerImportService;
+import org.hisp.dhis.tracker.imports.domain.TrackerObjects;
+import org.hisp.dhis.user.User;
+import org.hisp.dhis.user.UserService;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+
+class ChangeLogServiceTest extends TrackerTest {
+
+  @Autowired private ChangeLogService changeLogService;
+
+  @Autowired private TrackerImportService trackerImportService;
+
+  @Autowired private IdentifiableObjectManager manager;
+
+  @Autowired protected UserService _userService;
+
+  private User importUser;
+
+  private TrackerImportParams importParams;
+
+  @Override
+  protected void initTest() throws IOException {
+    userService = _userService;
+    setUpMetadata("tracker/simple_metadata.json");
+    importUser = userService.getUser("M5zQapPyTZI");
+    importParams = TrackerImportParams.builder().userId(importUser.getUid()).build();
+    assertNoErrors(
+        trackerImportService.importTracker(
+            importParams, fromJson("tracker/event_and_enrollment.json")));
+  }
+
+  @Test
+  void shouldFailWhenEventDoesNotExist() {
+    assertThrows(
+        NotFoundException.class,
+        () -> changeLogService.getEventChangeLog(UID.of(CodeGenerator.generateUid())));
+  }
+
+  @Test
+  void shouldFailWhenEventOrgUnitIsNotAccessible() {
+    injectSecurityContextUser(manager.get(User.class, "o1HMTIzBGo7"));
+
+    assertThrows(
+        NotFoundException.class, () -> changeLogService.getEventChangeLog(UID.of("D9PbzJY8bJM")));
+  }
+
+  @Test
+  void shouldFailWhenEventProgramIsNotAccessible() {
+    injectSecurityContextUser(manager.get(User.class, "o1HMTIzBGo7"));
+
+    assertThrows(
+        NotFoundException.class, () -> changeLogService.getEventChangeLog(UID.of("G9PbzJY8bJG")));
+  }
+
+  @Test
+  void shouldFailWhenProgramTrackedEntityTypeIsNotAccessible() {
+    injectSecurityContextUser(manager.get(User.class, "FIgVWzUCkpw"));
+
+    assertThrows(
+        NotFoundException.class, () -> changeLogService.getEventChangeLog(UID.of("H0PbzJY8bJG")));
+  }
+
+  @Test
+  void shouldFailWhenProgramWithoutRegistrationAndNoAccessToEventOrgUnit() {
+    injectSecurityContextUser(manager.get(User.class, "o1HMTIzBGo7"));
+
+    assertThrows(
+        NotFoundException.class, () -> changeLogService.getEventChangeLog(UID.of("G9PbzJY8bJG")));
+  }
+
+  @Test
+  void shouldReturnChangeLogsWhenDataValueIsCreated() throws NotFoundException {
+    injectSecurityContextUser(manager.get(User.class, "M5zQapPyTZI"));
+
+    Event event = manager.get(Event.class, "QRYjLTiJTrA");
+    assertNotNull(event);
+    String dataElementUid = event.getEventDataValues().iterator().next().getDataElement();
+    DataElement dataElement = manager.get(DataElement.class, dataElementUid);
+    assertNotNull(dataElement);
```

Comment:
I think we should also hide the annoyance of working with the set here 😅 
like in a `getFirstDataValue(event)`